File: Bi-Weekly Contest 178/Sum_of_GCD_of_Formed_Pairs.py

from math import gcd

class Solution:
    def gcdSum(self, nums: list[int]) -> int:
        prefixGcd = []
        curr_max = nums[0]

        for i in range(len(nums)):
            # Update the current maximum and calculate the GCD with it
            curr_max = max(curr_max, nums[i])
            curr_gcd = gcd(nums[i], curr_max)
            prefixGcd.append(curr_gcd)
        
        # Sort the prefix GCDs and use two pointers to calculate the sum of GCDs of pairs
        prefixGcd.sort()
        l = 0
        r = len(prefixGcd) - 1
        sums = 0
        
        while l < r:
            smallest = prefixGcd[l]
            largest = prefixGcd[r]
            # Add the GCD of the smallest and largest to the sum
            sums += gcd(smallest,largest)
            l += 1
            r -= 1

        return sums


sol = Solution()
inputs = [[2,6,4], [3,6,2,8]] # 2, 5
for nums in inputs:
    print(f"Input: nums = {nums}")
    print(f"Output: {sol.gcdSum(nums)}\n")

File: TypeScript/SumofGCDofFormedPairs.ts
function gcdSum(nums: number[]): number {
    let prefixGCD : number[] = [];
    var currentMax = nums[0];

    // Build the prefixGCD array.
    for (let i = 0; i < nums.length; i++){
        currentMax = Math.max(currentMax,nums[i]);
        prefixGCD.push(getGCD(nums[i], currentMax));
    }

    // Sort to pair the smallest and largest values.
    prefixGCD.sort((a, b) => a - b);

    var left = 0;
    var right = prefixGCD.length - 1;

    var sum = 0;

    // Sum the GCD of each smallest-largest pair.
    while (left < right){
        sum += getGCD(prefixGCD[left],prefixGCD[right]);
        left++;
        right--;
    }

    return sum;

};

function getGCD(a: number, b:number) : number{
    // Euclidean algorithm.
    if (b === 0) return a;
    return getGCD(b, a%b);
}


// Test Cases
console.log(gcdSum([2,6,4])); // Output: 2
console.log(gcdSum([3,6,2,8])); // Output: 5
console.log(gcdSum([5,6,14])); // Output: 1